File: src/main/kotlin/io/titandata/titan/providers/local/Tag.kt
/*
 * Copyright (c) 2019 by Delphix. All rights reserved.
 */

package io.titandata.titan.providers.local

import io.titandata.client.apis.CommitsApi

class Tag (
    private val commitsApi: CommitsApi = CommitsApi()
) {
    fun tagCommit(repository: String, commit: String, tags: List<String>) {
        val commit = commitsApi.getCommit(repository, commit)
        val commitTags = if (commit.properties.containsKey("tags")) {
            (commit.properties.get("tags") as Map<String, String>).toMutableMap()
        } else {
            mutableMapOf()
        }

        for (t in tags) {
            if (t.contains("=")) {
                val key = t.substringBefore("=")
                val value = t.substringAfter("=")
                commitTags[key] = value
            } else {
                commitTags[t] = ""
            }
        }

        val metadata = commit.properties.toMutableMap()
        metadata["tags"] = commitTags

        commitsApi.updateCommit(repository, io.titandata.models.Commit(id=commit.id, properties=metadata))
    }
}

File: internal/app/providers/common/Tag.go
package common

import (
	titanclient "github.com/titan-data/titan-client-go"
	"strconv"
	"strings"
)

func TagCommit(repo string, guid string, tags []string, port int) {
	cfg.BasePath = "http://localhost:" + strconv.Itoa(port)

	commit, _, _ := commitsApi.GetCommit(ctx, repo, guid)
	commitTags := make(map[string]string)
	t, ok := commit.Properties["tags"]
	if ok {
		commitTags = t.(map[string]string)
	}
	for _, tag := range tags {
		if strings.Contains(tag, "=") {
			s := strings.Split(tag, "=")
			commitTags[s[0]] = s[1]
		} else {
			commitTags[tag] = ""
		}
	}
	m := make(map[string]interface{})
	m = commit.Properties
	m["tags"] = commitTags
	c := titanclient.Commit{
		Id:         commit.Id,
		Properties: m,
	}
	commitsApi.UpdateCommit(ctx, repo, commit.Id, c)
}
